modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/LinkOrButton.js

```diff
@@ -7,54 +7,119 @@ import ClayButton from '@clayui/button';
-	const Wrapper = href && !disabled ? ClayLink : ClayButton;
+import React, {useEffect, useMemo, useState} from 'react';
 
-	return (
-		<>
-			<Wrapper
-				aria-label={symbol && ariaLabel}
-				block={otherProps.button?.block}
-				className={classNames(className, {
-					'd-md-none': responsive,
-					'nav-btn-monospaced': responsive,
-					'pl-4 pr-4': wide && !symbol,
-				})}
-				disabled={disabled}
-				href={href}
-				{...otherProps}
-				title={symbol && title}
+import './LinkOrButton.scss';
+
+const DEFAULT_BREAKPOINT_MD = '768px';
+
+/**
+ * A dynamic link/button that includes a symbol, a label, or both of them. All
+ * props are optional except for the label prop.
+ *
+ * It renders a ClayButton or a ClayLink depending on the provided:
+ * - disabled state (a ClayButton will be used).
+ * - href (a ClayLink will be used).
+ * - otherwise a ClayButton will be used.
+ *
+ * Then it shows a symbol on smaller screens, and label on larger screens.
+ * When only the symbol is being shown, an aria-label and title are added to the
+ * button for accessibility purposes. If there is no provided title, the label
+ * prop will be reused as title.
+ *
+ * It will always add a title to the button if it is specified as prop.
+ */
+function LinkOrButton(
+	{
+		className = '',
+		disabled = false,
+		href = '',
+		label,
+		symbol = '',
+		title = '',
+		...otherProps
+	},
+	ref
+) {
+	const [element, setElement] = useState(null);
+	const TagName = href && !disabled ? ClayLink : ClayButton;
+
+	const breakpointMd = element
+		? window.getComputedStyle(element).getPropertyValue('--breakpoint-md')
+		: DEFAULT_BREAKPOINT_MD;
+
+	const showLabel = useMediaQuery(`(min-width: ${breakpointMd})`);
+
+	const buttonProps = {
+		...otherProps,
+		className: 'link-or-button',
+		disabled,
+		href,
+		ref: mergeRefs(setElement, ref),
+	};
+
+	if (showLabel || !symbol) {
+		return (
+			<TagName
+				{...buttonProps}
+				className={classNames(
+					className,
+					buttonProps.className,
+					'link-or-button--with-label'
+				)}
+				title={title}
 			>
-				{symbol ? <ClayIcon symbol={symbol} /> : children}
-			</Wrapper>
-
-			{responsive && (
-				<Wrapper
-					block={otherProps.button?.block}
-					className={classNames(className, 'd-md-flex d-none', {
-						'pl-4 pr-4': wide,
-					})}
-					disabled={disabled}
-					href={href}
-					{...otherProps}
-				>
-					{children}
-				</Wrapper>
+				{label}
+			</TagName>
+		);
+	}
+
+	return (
+		<TagName
+			{...buttonProps}
+			aria-label={label}
+			className={classNames(
+				className,
+				buttonProps.className,
+				'nav-btn-monospaced'
 			)}
-		</>
+			title={title || label}
+		>
+			<ClayIcon symbol={symbol} />
+		</TagName>
 	);
-};
-export default LinkOrButton;
+}
+
+function mergeRefs(...refs) {
+	return (element) => {
+		refs.forEach((ref) => {
+			if (typeof ref === 'function') {
+				ref(element);
+			}
+			else if (ref) {
+				ref.current = element;
+			}
+		});
+	};
+}
```

Comment:
There are two cases here:

1. He need to keep the element inside a state because otherwise the component won't rerender when the ref is updated by react. We need to wait for the first render, then access it to get the CSS variable. There are some other ways to do this:
A. We could pass a function inside the ref, then build the media query inside this fn, but this still needs an extra state to call the other hook.
B. We could move all this logic inside the useMediaQuery hook, and access the CSS prop inside an useEffect (which will be run after the first render). I didn't go for this because I wanted to separate the "we need to get this CSS variable to build the query" and "we need to watch this query".

Personally I like using the useState because is a bit more explicit saying that "we do need to rerender when the ref is updated", but if we want to be precise we could build a hook called "run this just after first render", because the CSS variable is not going to change. We can even just embed the "768px" value here and go away, but it felt less future-proof.

2. I found that the forwardRef is needed if we are going to use the Component inside a ClayDropdown. It does need the ref to position the dropdown list, otherwise it is always on the bottom-left of the screen. I didn't notice until I deployed it without the forwardRef 😅

Reply:
> I found that the forwardRef is needed if we are going to use the Component inside a ClayDropdown. It does need the ref to position the dropdown list, otherwise it is always on the bottom-left of the screen. I didn't notice until I deployed it without the forwardRef

Doesn't this work in `master`, without `forwardRef`?

> the component won't re-render when the ref is updated by react

Do you mean updated by the parent component? `ref` does not seem to me like the right tool for that goal. It is suppose to stay stable over renders, working on separate lifecycle. If parent needs to control something in a child, the way do do it is simple prop. If it's position, then something like `position` props, or `otherProps` that allow setting of styles.

Reply:
> Doesn't this work in master, without forwardRef?

In master we are passing a regular ClayButton to the Dropdown :sweat_smile: 

> the component won't re-render when the ref is updated by react

I mean this:

1. LinkOrButton is called, ref is initialized to null, so the media query cannot be used.
2. LinkOrButton finishes rendering, ref.current is stablished to "HTMLButtonElement".
End :rainbow: 

If we do not use the ref inside a useEffect, we will never get a run where ref.current has some value, using a state we force the component to update after the first render.

Reply:
And again, we can put the `768px` in the JS and call it a day, I don't think that value is going to change anytime soon.

Reply:
> In master we are passing a regular ClayButton to the Dropdown

Ah, right. Yeah, if `ClayDropdown` needs the ref, for whatever reason, we need to provide it :+1: 

> LinkOrButton is called, ref is initialized to null, so the media query cannot be used

I am probably missing something here, isn't media query unrelated to refs? Is that the root reason for the element in state? `showLabel` is just a local variable, not a state variable, so what triggers the component update?  Does media query trigger ref update, is that what we're latching onto?

I'm confused because normally, just for the sake of `forwardRef`, all this element state shenanigans in not necessary, like for example in [ClayButton](https://github.com/liferay/clay/blob/master/packages/clay-button/src/Button.tsx#L145).

Reply:
Anyways, I don't want to block this for long. If this works, we can push it, and then later on possibly simplify it.

Reply:
@markocikos I've been thinking about this the whole weekend... :sweat_smile: 

Basically I think that the ref to get the value of the custom prop added too much complexity, so I added a new commit using just the constant value instead of getting it from the ref. This simplify the code a lot and just leaves the useMediaQuery hook.

Reply:
I added this in a separated commit for the case you want to have a look.

Reply:
Nice, that's much cleaner! :sparkles: